community/values/src/main/java/org/neo4j/values/virtual/VirtualValues.java

```diff
@@ -39,12 +49,103 @@ private VirtualValues()
 
     public static ListValue list( AnyValue... values )
     {
-        return new ListValue( values );
+        return new ListValue.ArrayListValue( values );
+    }
+
+    public static ListValue fromList( List<AnyValue> values )
+    {
+        return new ListValue.JavaListListValue( values );
+    }
+
+    public static ListValue range( long start, long end, long step )
+    {
+        return new ListValue.IntegralRangeListValue( start, end, step );
+    }
+
+    public static ListValue fromArray( ArrayValue arrayValue )
+    {
+        return new ListValue.ArrayValueListValue( arrayValue );
+    }
+
+    public static ListValue filter( ListValue list, Function<AnyValue,Boolean> filter )
+    {
+        return new ListValue.FilteredListValue( list, filter );
+    }
+
+    public static ListValue slice( ListValue list, int from, int to )
+    {
+        int f = Math.max( from, 0 );
+        int t = Math.min( to, list.size() );
+        if ( f > t )
+        {
+            return EMPTY_LIST;
+        }
+        else
+        {
+            return new ListValue.ListSlice( list, f, t );
+        }
+    }
+
+    public static ListValue drop( ListValue list, int n )
+    {
+        int start = Math.max( 0, Math.min( n, list.size() ) );
+        return new ListValue.ListSlice( list, start, list.size() );
+    }
+
+    public static ListValue take( ListValue list, int n )
+    {
+        int end = Math.max( 0, Math.min( n, list.size() ) );
+        return new ListValue.ListSlice( list, 0, end );
+    }
+
+    public static ListValue transform( ListValue list, Function<AnyValue,AnyValue> transForm )
+    {
+        return new ListValue.TransformedListValue( list, transForm );
+    }
+
+    public static ListValue reverse( ListValue list )
+    {
+        return new ListValue.ReversedList( list );
```

Comment:
I like all these functional methods, but we to make a few unit tests to verify such central functionality.